realm/realm-library/src/androidTest/java/io/realm/rule/RunInLooperThread.java

```diff
@@ -229,11 +265,177 @@ public void postRunnableDelayed(Runnable runnable, long delayMillis) {
     public void looperTearDown() {
     }
 
+    private void initRealm() {
+        synchronized (lock) {
+            realm = Realm.getInstance(realmConfiguration);
+        }
+    }
+
+    private void closeRealms() {
+        closeTestRealms();
+
+        Realm oldRealm;
+        synchronized (lock) {
+            oldRealm = realm;
+
+            realm = null;
+            realmConfiguration = null;
+        }
+
+        if (oldRealm != null) {
+            oldRealm.close();
+        }
+    }
+
     /**
      * If an implementation of this is supplied with the annotation, the {@link RunnableBefore#run(RealmConfiguration)}
      * will be executed before the looper thread starts. It is normally for populating the Realm before the test.
      */
     public interface RunnableBefore {
         void run(RealmConfiguration realmConfig);
     }
+
+    class RunInLooperThreadStatement extends Statement {
+        private final RunTestInLooperThread annotation;
+        private final Statement base;
+
+        RunInLooperThreadStatement(RunTestInLooperThread annotation, Statement base) {
+            this.annotation = annotation;
+            this.base = base;
+        }
+
+        @Override
+        @SuppressWarnings({"ClassNewInstance", "Finally"})
+        public void evaluate() throws Throwable {
+            before();
+
+            Class<? extends RunnableBefore> runnableBefore = annotation.before();
+            if (!runnableBefore.isInterface()) {
+                // this is dangerous: config is mutable.
+                runnableBefore.newInstance().run(getConfiguration());
+            }
+
+            runTest(annotation.threadName());
+        }
+
+        @SuppressWarnings({"unused", "ThrowFromFinallyBlock"})
+        private void runTest(final String threadName) throws Throwable {
+            Exception testException = null;
+
+            try {
+                ExecutorService executorService = Executors.newSingleThreadExecutor(new ThreadFactory() {
+                    @Override
+                    public Thread newThread(Runnable runnable) { return new Thread(runnable, threadName); }
+                });
+
+                TestThread test = new TestThread(base);
+
+                Future<?> ignored = executorService.submit(test);
+
+                TestHelper.exitOrThrow(executorService, signalTestCompleted, test);
+            } catch (Exception error) {
+                // These exceptions should only come from TestHelper.awaitOrFail()
+                testException = error;
+            } finally {
+                // Tries as hard as possible to close down gracefully, while still keeping all exceptions intact.
+                cleanUp(testException);
+
+                // Only TestHelper.awaitOrFail() threw an exception
```

Comment:
Any reason this block of code isn't inside `cleanup` it looks slightly weird to me right now?